### ISSUE TITLE
MacOS system curl broken after letsencrypt CA cert expiry

### ISSUE THREAD

**Issue by wwuck** [Homebrew/brew#12161]
### `brew config` output

```shell
HOMEBREW_VERSION: 3.2.14
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4e6919b73444e9f9e02ad81d8afc4bbd97533567
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c437fbe02752c09d3bad41eb2ab84832ea15968f
Core tap last commit: 20 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit kabylake
Clang: 11.0.0 build 1100
Git: 2.33.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 11.3.1
```

### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/more/.
You should download the Command Line Tools for Xcode 11.3.1.


Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> /Users/me/.bash_profile
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I was trying to reinstall `openconnect` from HEAD using `brew reinstall openconnect`.
I had originally installed `openconnect` using `brew install openconnect --HEAD`.

### What happened (include all command output)?

```console
$ brew reinstall openconnect
==> Downloading https://git.infradead.org/users/dwmw2/vpnc-scripts.git/blob_plain/c0122e891f7e033f35f047dad963702199d5cb9e:/vpnc-script

curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
HTTPS-proxy has similar options --proxy-cacert and --proxy-insecure.
Error: Failed to download resource "openconnect--vpnc-script"
Download failed: https://git.infradead.org/users/dwmw2/vpnc-scripts.git/blob_plain/c0122e891f7e033f35f047dad963702199d5cb9e:/vpnc-script
```

### What did you expect to happen?

`openconnect` should be installed from latest `HEAD`.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew reinstall openconnect
```

**Comment by wwuck** [comment]
I believe this is due to the expiry of a LetsEncrypt CA certificate.

https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/

If I try to download the openconnect file using system curl, it fails.
```console
$ /usr/bin/curl https://git.infradead.org/users/dwmw2/vpnc-scripts.git/blob_plain/c0122e891f7e033f35f047dad963702199d5cb9e:/vpnc-script
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
HTTPS-proxy has similar options --proxy-cacert and --proxy-insecure.
```

If I use brew-installed curl, then it succeeds.
```console
$ /usr/local/opt/curl/bin/curl https://git.infradead.org/users/dwmw2/vpnc-scripts.git/blob_plain/c0122e891f7e033f35f047dad963702199d5cb9e:/vpnc-script
#!/bin/sh
#
# Originally part of vpnc source code:
# © 2005-2012 Maurice Massar, Jörg Mayer, Antonio Borneo et al.
# © 2009-2012 David Woodhouse <dwmw2@infradead.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
################
#
# List of parameters passed through environment
...
```
Homebrew should be forcing the use of homebrew-installed curl for `brew install` on systems that don't have a system curl compatible with the new LetsEncrypt root CA certificates.

**Comment by carlocab** [comment]
Thanks for the report -- we've encountered this problem in CI in quite a few PRs already.

In the meantime, as a workaround, you can set `HOMEBREW_FORCE_BREWED_CURL=1`.

**Comment by Bo98** [comment]
We're looking into this but it's tricky. For example, you need `gettext` to install brewed curl, but to install `gettext` you need brewed curl (unless you can use a bottle).

**Comment by SMillerDev** [comment]
Is there a way we can just pull in (or point to) the new Root for Letsencrypt?

**Comment by FiloSottile** [comment]
This seems to also affect custom Taps that have Let's Encrypt chains.

https://twitter.com/brunobord/status/1444053423461130244

**Comment by Bo98** [comment]
Thanks for the heads up about that! We might need to enforce brewed git for that too then (`HOMEBREW_FORCE_BREWED_GIT` is the workaround for that in the meantime).

**Comment by FiloSottile** [comment]
FWIW, I can't reproduce neither the git nor the curl issue on macOS 11. My best guess is that those tools on macOS 10.14 have a chain building issue like or due to this https://www.openssl.org/blog/blog/2021/09/13/LetsEncryptRootCertExpire/.

**Comment by Bo98** [comment]
> due to this https://www.openssl.org/blog/blog/2021/09/13/LetsEncryptRootCertExpire/.

Yes it's exactly that. In terms of LibreSSL (what Apple use for curl), it was an bug up to LibreSSL 3.2.0. But Apple did backport a fix to 2.8.3 in macOS 10.15.6.

So the bug affects 10.13-10.15.5 (curl uses SecureTransport before 10.13).
10.10 and 10.11 are also affected because they don't contain the ISRG Root X1 certificate in their default cert store at all.
10.12 weirdly enough should work fine as far as I know.

**Comment by carlocab** [comment]
Yes, this definitely affects only Mojave and earlier. I believe it's a macOS LibreSSL bug.

**Comment by peterdd** [comment]
On old macbook pro using 10.13 High Sierra:

```
/opt/local/bin/curl -V
curl 7.62.0 (x86_64-apple-darwin17.7.0) libcurl/7.62.0 OpenSSL/1.0.2p zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5)
Release-Date: 2018-10-31
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy PSL 

/usr/local/opt/curl/bin/curl -V
curl 7.79.1 (x86_64-apple-darwin17.7.0) libcurl/7.79.1 (SecureTransport) OpenSSL/1.1.1l zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 libidn2/2.3.2 libssh2/1.10.0 nghttp2/1.45.1 librtmp/2.3 OpenLDAP/2.5.7
Release-Date: 2021-09-22
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets zstd
```

seems both using OpenSSL, not LibreSSL?

Also the suggested

`echo 'export PATH="/usr/local/opt/curl/bin:$PATH"' >> /Users/yourusername/.bash_profile`

and opening a new terminal seemed to not help with the `brew upgrade` command.

**Comment by Bo98** [comment]
> seems both using OpenSSL, not LibreSSL?

Neither of of the curl versions you've shown is from Apple. Homebrew's version (bottom) uses OpenSSL 1.1, which is not affected by the bug.

> Also the suggested
> 
> `echo 'export PATH="/usr/local/opt/curl/bin:$PATH"' >> /Users/yourusername/.bash_profile`
> 
> and opening a new terminal seemed to not help with the `brew upgrade` command.

The suggestion above is `HOMEBREW_FORCE_BREWED_CURL=1`.

**Comment by peterdd** [comment]
Thanks @Bo98, I had this in .bash_profile:

```
# MacPorts Installer addition on 2018-11-15_at_20:13:58: adding an appropriate PATH variable for use with MacPorts.
export PATH="/opt/local/bin:/opt/local/sbin:$PATH"
# Finished adapting your PATH environment variable for use with MacPorts.

export PATH="/usr/local/opt/curl/bin:$PATH"
```

With disabling the 2 exports there I got:

```
curl -V
curl 7.54.0 (x86_64-apple-darwin17.0) libcurl/7.54.0 LibreSSL/2.0.20 zlib/1.2.11 nghttp2/1.24.0
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz HTTP2 UnixSockets HTTPS-proxy
```

Using

```
export HOMEBREW_FORCE_BREWED_CURL=1
brew upgrade
```

I was able to upgrade. :-D

**Comment by pukku** [comment]
Is there a solutions for those of us on 10.14 that don't already have a brewed curl? I can't install the brew version of curl, because of the problem, but without installing the brew version of curl, I can't fix the problem?

**Comment by EricFromCanada** [comment]
This [answer at Stack Overflow](https://stackoverflow.com/a/69396425) shows which entry to delete from `/etc/ssl/cert.pem` to get things working on your machine.

**Comment by SMillerDev** [comment]
As far as I know you should just be able to install the bottles Homebrew provides as well, it should only be building from source that's broken.

**Comment by pukku** [comment]
I don't see a way to install a bottle downloaded myself? The bottles don't download via brew install because of (I think?) the curl problem?

But I deleted the certificate from the /etc/ssl/cert.pem file (after backing it up), and that at least seems to let me install curl for source (and all the prerequisites from source as well, apparently) — at least, so far (it's still working)...

Thanks!

**Comment by Bo98** [comment]
Bottles should work fine. If not, post the logs and I'll have a look.

The hope is to release 3.2.15 with an automatic fix for this soon. We've restructured a bit how some of our formulae work but it's getting close to completion now. I've got both a 10.10 and 10.14 install to test this on as well.

**Comment by pukku** [comment]
I think the problem was that I hadn't updated brew in a while, and it wouldn't let me download something that was needed to update? At any rate, after I removed the certificate from the pem file, I was able to run brew update, and then bottles started working again. At any rate, everything is fine now. Many thanks!

**Comment by Bo98** [comment]
The current plan is:

10.10-10.11 - install an up-to-date curl and git. We will make sure curl is installable even with the broken system curl, and then use the brewed curl to install anything else (including git).
10.12-10.15.5 - install our own CA store and use that for system curl and git usages within brew instead of the system. People can use this outside of brew with `SSL_CERT_FILE` and `GIT_SSL_CAINFO` if they wish to use system curl and git, or they can install updated versions of curl and git from us which will work by default.

These changes will be a part of 3.2.15, which I'm trying to get out as soon as possible. Most changes are already in place in `master` - we're just working on some final fixes.

**Comment by Bo98** [comment]
We've released [3.2.15](https://github.com/Homebrew/brew/releases/tag/3.2.15) with a full fix for this.

Short summary: You don't need to set any envs or do any workarounds anymore (though you can keep them if you want) - it should all work automatically now.

Long changelog:

* `ca-certificates` has been added to homebrew-core (https://github.com/Homebrew/homebrew-core/pull/86304)
  - Brew will now install this automatically when running macOS <= 10.15.5 (https://github.com/Homebrew/brew/pull/12167), and will use it instead of the system CA store when working with system curl and git. This behaviour can be forced on other platforms (like Linux) by `HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1`.
    * If building from source, we make sure even a curl with broken certificates can still install this (https://github.com/Homebrew/brew/pull/12172).
  - For Yosemite and El Capitan, a full install of Homebrew curl and git are required and will be performed automatically.
  - If you are using our `openssl@1.1`, our CA store used by default (other TLS libraries coming soon), but it can also be used externally should whatever you are using allow swaping CA certificate stores.
* We've made the installation of Homebrew's curl easier
  - When building from source (i.e. not using bottles), plain HTTP mirrors have been added (https://github.com/Homebrew/homebrew-core/pull/86292, https://github.com/Homebrew/homebrew-core/pull/86293, https://github.com/Homebrew/homebrew-core/pull/86294, https://github.com/Homebrew/homebrew-core/pull/86295, https://github.com/Homebrew/homebrew-core/pull/86296, https://github.com/Homebrew/homebrew-core/pull/86297, https://github.com/Homebrew/homebrew-core/pull/86298, https://github.com/Homebrew/homebrew-core/pull/86299, https://github.com/Homebrew/homebrew-core/pull/86300, https://github.com/Homebrew/homebrew-core/pull/86301)
    - We will make sure this continues to be offered (https://github.com/Homebrew/brew/pull/12166).
    - Note that on Linux, this doesn't cover the full dependency tree yet (but if you are experiencing SSL cert issues there then `HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1` should be enough)
  - Less dependencies are installed:
    * `cmake` no longer installs documentation - this is now in a separate formula `cmake-docs` (https://github.com/Homebrew/homebrew-core/pull/86249).
    * `c-ares` now only depends has one build-time dependency (https://github.com/Homebrew/homebrew-core/pull/86250/)
    * The libraries from `nghttp2` have been split into a separate formula (https://github.com/Homebrew/homebrew-core/pull/86291).
* Outside of brew's core operations, we're fixing formula-specific issues on a case-by-case basis. If you spot a SSL certificate error with a specific formula, open a homebrew-core issue and let us know.
  - Homebrew's git now uses our own libcurl on <= Mojave (https://github.com/Homebrew/homebrew-core/pull/86439) so if you currently using system git and are experiencing issues, you can either switch to our git. Alternatively, you can swap out the default SSL store used by system git to point to our CA store (via `GIT_SSL_CAINFO`), which is installed by default on affected macOS systems.

There's more but I think that covers most of noticeable things.
